stdio/file.c

```diff
@@ -311,98 +320,186 @@ static int full_read(int fd, void *ptr, size_t size)
 }
 
 
-static int full_write(int fd, const void *ptr, size_t size)
+static ssize_t full_write(int fd, const void *ptr, size_t size)
 {
-	int err;
-	while (size) {
-		if ((err = __safe_write(fd, ptr, size)) < 0) {
-			return -1;
+	ssize_t err;
+	ssize_t total = 0;
+
+	while (size > 0) {
+		err = __safe_write_nb(fd, ptr, size);
+		if (err < 0) {
+			return (errno == EAGAIN) ? total : -1;
 		}
 		ptr += err;
+		total += err;
 		size -= err;
 	}
 
-	return size;
+	return total;
 }
 
 
-size_t fread_unlocked(void *ptr, size_t size, size_t nmemb, FILE *stream)
+static int __fflush_one(FILE *stream)
 {
-	int err;
-	size_t readsz = nmemb * size;
-	size_t bytes;
+	int ret = 0;
+	ssize_t err;
+	off_t off;
 
-	if (!readsz) {
+	if (stream->buffer == NULL) {
 		return 0;
 	}
 
-	if (stream->buffer == NULL) {
-		/* unbuffered read */
-		if ((err = full_read(stream->fd, ptr, readsz)) < 0) {
-			return 0;
+	if ((stream->flags & F_WRITING) != 0) {
+		if (stream->bufpos != 0) {
+			err = full_write(stream->fd, stream->buffer, stream->bufpos);
+			if (err != stream->bufpos) {
+				stream->flags |= F_ERROR;
+				ret = -1;
+			}
+			else {
+				stream->bufpos = 0;
+			}
+		}
+	}
+	else {
+		if (stream->bufpos != stream->bufeof) {
+			off = lseek(stream->fd, (off_t)stream->bufpos - stream->bufeof, SEEK_CUR);
```

Comment:
Should this be as-is or the other way around?

```suggestion
			off = lseek(stream->fd, (off_t)stream->bufeof - stream->bufpos, SEEK_CUR);
```

Reply:
This is correct, we are seeking backwards: `-(bufeof - bufpos) = bufpos - bufeof`.